### PR TITLE
[Maxtext] [Tunix] New Weight Sync for Post-train, maxtext -> maxtext

### PR DESCRIPTION
**Pull request by YixuanWang-99**
# Description

This PR introduce a new weightConverter pipeline of the MaxText to vLLM weight conversion engine. Current use case is `MaxTextForCausalLM ` for model `Qwen30-30B-A3B`. The Maxtext side PR is: https://github.com/AI-Hypercomputer/maxtext/pull/4524

For now, the `WeightConverter` is optional through `use_weight_converter` flag from the `vllm_additional_config` JSON dictionary. With this `use_weight_converter: false`, the weight conversion falls back to original `transfer_state_directly()` behavior. 

# Tests

1. Unit tests, `maxtext/tests/post_training/unit/weight_converter_test.py`
2. Integration test, using legacy transfer_state_directly function call`validate_converter.py`, [logs](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-multipod-dev%22%0Aresource.labels.location%3D%22europe-west4%22%0Aresource.labels.cluster_name%3D%22mlperf-v5p%22%0Aresource.labels.pod_name:%22q35legacy-1786578447-%22%0Aresource.labels.container_name:%22jax-tpu%22%0Aseverity%3E%3DDEFAULT;cursorTimestamp=2026-08-12T23:54:03.861926240Z;customDuration=today?e=13802955&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev).
The conversion+reshard+assign time is [around 22.3s (25.34s total - 3.02 sharding in tpu v5p-8](https://pantheon.corp.google.com/logs/query;cursorTimestamp=2026-08-12T23:53:57.388785997Z;customDuration=today;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-multipod-dev%22%0Aresource.labels.location%3D%22europe-west4%22%0Aresource.labels.cluster_name%3D%22mlperf-v5p%22%0Aresource.labels.pod_name:%22q35legacy-1786578447-%22%0Aresource.labels.container_name:%22jax-tpu%22%0Aseverity%3E%3DDEFAULT%0Atimestamp%3D%222026-08-12T23:53:57.388678966Z%22%0AinsertId%3D%228wmar5tbj1pqp0gs%22?project=cloud-tpu-multipod-dev)

using new weightConverter with `use_weight_converter=true`, [logs](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-multipod-dev%22%0Aresource.labels.location%3D%22europe-west4%22%0Aresource.labels.cluster_name%3D%22mlperf-v5p%22%0Aresource.labels.pod_name:%22q35val-1786576710-%22%0Aresource.labels.container_name:%22jax-tpu%22%0Aseverity%3E%3DDEFAULT;cursorTimestamp=2026-08-12T23:24:36.181166892Z;customDuration=today?e=13802955&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev)
The conversion+reshard+assign time is [15.24s in v5p-8](https://pantheon.corp.google.com/logs/query;cursorTimestamp=2026-08-12T23:24:36.181166892Z;customDuration=today;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-multipod-dev%22%0Aresource.labels.location%3D%22europe-west4%22%0Aresource.labels.cluster_name%3D%22mlperf-v5p%22%0Aresource.labels.pod_name:%22q35val-1786576710-%22%0Aresource.labels.container_name:%22jax-tpu%22%0Aseverity%3E%3DDEFAULT%0Atimestamp%3D%222026-08-12T23:24:32.408123294Z%22%0AinsertId%3D%22fbi666f40w5ahxq1%22?project=cloud-tpu-multipod-dev)
```
python -m maxtext.integration.vllm.torchax_converter.validate_converter \
      src/maxtext/configs/post_train/rl.yml \
      model_name=qwen3-30b-a3b \
      tokenizer_type=huggingface \
      tokenizer_path=Qwen/Qwen3-30B-A3B \
      load_parameters_path='gs://runner-maxtext-logs/qwen3-30b-a3b-base/to_maxtext/scanned/2026-06-15-22-21-28/0/items' \
      run_name=validate_qwen35_maxtext_integration \
      vllm_hf_overrides=\"{'architectures': ['MaxTextForCausalLM']}\" \
      vllm_additional_config='{\"maxtext_config\": {\"model_name\": \"qwen3-30b-a3b\", \"model_call_mode\": \"inference\", \"enable_dp_attention\": false, \"allow_split_physical_axes\": true, \"log_config\": false, \"weight_dtype\": \"bfloat16\", \"prefuse_moe_weights\": true}, \"use_weight_converter\": true}' \
      use_standalone_converter=False \
      max_prefill_predict_length=8 \
      max_target_length=16 \
      scan_layers=true \
      skip_jax_distributed_system=true \
      weight_dtype=bfloat16 \
      rollout_tensor_parallelism=4 \
      hf_access_token=<token>\
      prompt=\"Paris is\""
```

4.End-to-End test, `train_rl`, [logs](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-multipod-dev%22%0Aresource.labels.location%3D%22europe-west4%22%0Aresource.labels.cluster_name%3D%22mlperf-v5p%22%0Aresource.labels.pod_name:%22q30rl-1786560347-%22%0Aresource.labels.container_name:%22jax-tpu%22%0Aseverity%3E%3DDEFAULT;cursorTimestamp=2026-08-12T19:05:03.550298696Z;customDuration=today?e=13802955&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev)
```
python -m maxtext.trainers.post_train.rl.train_rl \
      base_output_directory='gs://runner-maxtext-logs/weightConverter/qwen3-30b-a3b/0803' \
      load_parameters_path='gs://runner-maxtext-logs/qwen3-30b-a3b-base/to_maxtext/scanned/2026-06-15-22-21-28/0/items' \
      run_name='rl_test_08_03' \
      rl.loss_algo='grpo' \
      scan_layers=true \
      num_batches=5 \
      batch_size=16 \
      num_test_batches=5 \
      model_name=qwen3-30b-a3b \
      tokenizer_type=huggingface \
      tokenizer_path=Qwen/Qwen3-30B-A3B \
      enable_single_controller=true \
      checkpoint_storage_use_zarr3=False \
      checkpoint_storage_use_ocdbt=False \
      rollout_tensor_parallelism=4 \
      async_scheduling=false \
      vllm_hf_overrides=\"{'architectures': ['MaxTextForCausalLM']}\" \
      vllm_additional_config='{\"maxtext_config\": {\"model_name\": \"qwen3-30b-a3b\", \"enable_dp_attention\": false, \"allow_split_physical_axes\": true, \"log_config\": false, \"weight_dtype\": \"bfloat16\", \"prefuse_moe_weights\": true}, \"use_weight_converter\": true}'"
```
**Checklist**

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and all unit tests pass.
- [x] I have added all appropriate doc-strings/documentation.
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).
- [x] I have followed [Contribution Guidelines](https://github.com/google/tunix/blob/main/docs/contributing.md).

### CODE REVIEW

File: tunix/generate/vllm_sampler.py [hunk 1]
```diff
@@ -199,7 +201,49 @@ def update_params(
     # Synchronization point before weight sync
     jax.effects_barrier()
 
-    if self.to_hf_key_mappings:
+    if self.converter is not None:
+      from flax import traverse_util, nnx
+      logging.info("Using native WeightConverter specifically for MaxText integration.")
+      vllm_state = self.converter.convert(
```
Comment (YixuanWang-99) [inline tunix/generate/vllm_sampler.py]: separate weight conversion and resharding for easy use and better maintainance

File: tunix/generate/vllm_sampler.py [hunk 2]
```diff
@@ -43,6 +44,54 @@
...
+  """Reports whether the reshard actually moved weights onto the target mesh.
+
+  Conversion legitimately produces trainer-mesh arrays; the reshard is what
+  crosses onto the rollout mesh. If a resharded array still reports the source
+  mesh here, the crossing silently did not happen and the rollout will later
+  execute against buffers its own client cannot address.
+  """
+  stale = []
+  for key, value in resharded_flat.items():
+    want = _placement(spec_flat[key])
```
Comment (wang2yn84) [inline tunix/generate/vllm_sampler.py]: We mainly use _placement for comparison, do we really need to compare string? Compare 2 list should be fine.

File: tunix/generate/vllm_sampler.py
Comment (wang2yn84) [inline tunix/generate/vllm_sampler.py]: Any reason why we change the output type here?

Comment (YixuanWang-99) [reply tunix/generate/vllm_sampler.py]: vllm types it Sequence[] but util declares List[]. I casted it as list() to do len() checks later.

File: tunix/generate/vllm_sampler.py (hunk 2 above)
Comment (YixuanWang-99) [reply tunix/generate/vllm_sampler.py]: Good catch, I changed to compare two lists elementwise. Btw, comparing strings can be wrong becase the the middle of the reshared array can get omitted, I changed it compare every element instead.

Comment (YixuanWang-99) [reply tunix/generate/vllm_sampler.py]: Btw, moved the debugging helper functions into utils.py

File: tunix/generate/vllm_sampler.py
Comment (YixuanWang-99) [reply tunix/generate/vllm_sampler.py]: I saw changes in the main branch solved this issue. I will follow the main branch.
